### DISCUSSION TITLE
ATtiny816 not found on OneWire

### DISCUSSION THREAD

**Discussion by chris-b-tanner** [SpenceKonde/megaTinyCore#995] [Q&A]
Hi folks, I originally posted this as an issue but it might be better here.

I am struggling to get the ATtiny816 to respond as a OneWire device to an ESP32 master. Running the DS18B20 example.

Background info:
The bus and master physically work. I have another real DS18B20 on the same OneWire bus and it's seen fine by the host (an ESP32); the reset pulse from the ESP32 looks a strong low on a 'scope. I've tested the ATtiny without this separate device on the bus as well.
I can run the DS18B20 emulation code fine on an ATtiny85, so it's something related to the ATtiny816.
4k7 pullup is on the ESP32 end.
I'm using PlatformIO, but I don't suspect that makes a difference.
Nothing else is hooked up to the ATtiny apart from a 1uF cap from VCC to GND and the debug LED.
The DS18B20 emulation code runs on the ATtiny816, with the LED/blink at exactly 1s, but nothing on OneWire.
I've confirmed the OneWire pin I'm using on the ATtiny (10 / PB4) can be manipulated fine from the code, and it receives the low pulse into the calibration code.
The ATtiny816 runs the calibration code with a serial output, but I had to edit to **while (repetitions++ < 5)** in order for it to come out of the other end with a value_ipl (which strangely gives me 11 not 13 which is the default for the AVR)
No response on OneWire for IPL at either 11, 12 or 13

I have set fuses to use both 16Mhz and 20Mhz internal clock.

OneWireHub version 3.0.0
PlatformIO config:

```[env:attiny816]
platform = atmelmegaavr
platform_packages = platformio/tool-avrdude
board = attiny816
build_flags = -DCLOCK_SOURCE=6
framework = arduino
board_build.f_cpu = 16000000L
upload_protocol = serialupdi
lib_deps = orgua/OneWireHub

**Comment by chris-b-tanner** [comment]
Oh, shoot, I posted this in the wrong repo, sorry - it was supposed to be over at OneWireHub. If I get a solution I'll update it here anyway in case it's useful.

**Comment by chris-b-tanner** [comment]
Actually, I'll reopen this as I'm wondering if it's because of differences between the atmelavr and atmelmegaavr platforms as compiled into the final code. Might there be something in the megaTinyCore which is causing trouble?

**Reply by SpenceKonde** [reply]
Which OneWire is this? 
https://github.com/PaulStoffregen/OneWire ?
That one that special cases the ATmega4809, and **intentionally chooses not to support any other modern AVR**. There are a dozen PR's open and ignored that, some dating back years. The author has decided that he will not release a version of the library which supports these parts, and that is his prerogative. So I forked that obstructionist's library and applied the one bloody trivial fix, as this is open source that that is my perogative....  I was sorta hoping it would got merged in and all but - nope. All he supports is the 4809. The literal exact same code works for every modern AVR (though iirc there were possible optimizations) https://github.com/SpenceKonde/OneWire

**Reply by chris-b-tanner** [reply]
Thanks Spence - I have tried both those libraries (on the master, an ESP32 note) - but neither can find the emulated slave when OneWireHub is running on an Attiny816. The exact same emulation code (bar the mega core) is running fine on an ATtiny85, which is found by the ESP32 with both your and Paul's libraries. Which makes me think there's something in the mega core which doesn't play nice with OneWireHub on the emulated device.

**Reply by chris-b-tanner** [reply]
Perhaps a clue: I have looked at the package usage in PlatformIO and it's pulling in 
framework-arduino-megaavr-megatinycore @ ^2.6.7

But I can't update it using plaformio.ini 

```
Resolving attiny816 dependencies...
Tool Manager: Installing platformio/framework-arduino-megaavr-megatinycore @ ^2.6.8
UnknownPackageError: Could not find the package with 'platformio/framework-arduino-megaavr-megatinycore @ ^2.6.8' requirements for your system 'darwin_x86_64'
```


Interestingly your notes on the 2.6.8 release tag suggest that 2.6.7 won't compile but I've not had problems.
UPDATE: this was a red herring. I've repeated a load of test steps using ArduinoIDE (megatinycore 2.6.8) instead of PlatformIO and all problems appear the same.

**Comment by SpenceKonde** [comment]
That's unfortunate. I thibk the cores need to be manually updated past that
on PIO. But that shouldn't be related to the issue.

And you didn't answer the one most important question: there are MANNNYYYY
libraries called OneWire, which one are you using???

Like I said the most popular library of thst name maintained by Paul S
appears to have made the deliberate decision not to support these parts and
his version of the library is unsuitable for these parts and since he
refused to merge my fixes or any of the other dozen plus other prs to fix
it, I forked it and the fork should work

____________
Spence Konde
Azzy’S Electronics

New products! Check them out at tindie.com/stores/DrAzzy
GitHub: github.com/SpenceKonde
ATTinyCore: Arduino support for almost every ATTiny microcontroller
Contact: ***@***.***

**Reply by chris-b-tanner** [reply]
Hi Spence, I did answer ;) - tried both Paul's and yours - but I'm not sure that's relevant as the OneWire library is running on the master, which is an ESP32. 

I'm wondering if the ATtiny816 might be running at a different loop speed from the ATtiny85, even though both have fuses set to 16MHz internal and both appear to be doing timed things properly (eg a 1 second tick is exactly one second on both). Don't know if this is related to the megaTinyCore vs the old avr core. I'll try and play around with a 'scope to pinpoint the problem.
